### PR TITLE
[CONTINT-3683] add webhook to inject default agent sidecar

### PR DESCRIPTION
**Pull request by adel121**
**ATTENTION**: This PR is better reviewed commit-by-commit.

### What does this PR do?

This PR creates and registers a new webhook that allows automatic agent sidecar injection into users applicative pods based on namespace label selector and pod label selector user configuration.

### Motivation

Facilitate agent setup in environments where agent should be launched as a sidecar instead using a daemonset (e.g. EKS Fargate)

### Possible Drawbacks / Trade-offs

- With the current implementation, the agent sidecar will be injected, but will fail to run on EKS Fargate. The reason for this is that `DD_EKS_FARGATE` label is not set by default, in order to allow a more generic solution. In a future PR, we will add the capability for the user to add ENV var overrides in order to customise the injected agent sidecar.
- With the current implementation, the agent can't communicate with the cluster agent. This can be added in a future PR if needed.
- With this solution, we assume the user has created a datadog secret in every namespace where they which to inject the agent sidecar
- If the user doesn't set any selectors (or we fail to parse the selectors set by the user), the webhook will end up receiving and processing pod mutation requests for all pods in all namespaces. (Should we prevent this behaviour ❓ )

### Describe how to test/QA your changes

Testing on kind (you could do the same on EKS Fargate, but beware that you should create fargate profiles in order to run pods in fargate): 

- Deploy the clusteragent (without the agent) on a kind cluster with the following:
   - Admission controller enabled
   - Agent sidecar injection enabled
   - Agent sidecar selectors set
   - Agents disabled

#### Basic Configuration (For fargate)
```
datadog:
  apiKeyExistingSecret: datadog-secret
  appKeyExistingSecret: datadog-secret
  kubelet:
    tlsVerify: false

agents:
  enabled: false

clusterAgent:
  enabled: true
  admissionController:
    enabled: true
  env:
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_ENABLED
      value: "true"
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROVIDER
      value: "fargate"
```

With this configuration, only pod have the `"agent.datadoghq.com/sidecar": "fargate"` label will be injected with the agent sidecar, and the `DD_EKS_FARGATE` label should be set to `true` on the agent sidecar container.

Create two pods:
- Pod with label:
```
apiVersion: v1
kind: Pod
metadata:
  name: pod-with-label
  labels:
    admission.datadoghq.com/enabled: "false"
    agent.datadoghq.com/sidecar: "fargate"
spec:
  containers:
    - name: nginx
      image: nginx:1.14.2
      ports:
        - containerPort: 80
```
- Pod without label:
```
apiVersion: v1
kind: Pod
metadata:
  name: pod-without-label
  labels:
    admission.datadoghq.com/enabled: "false"
spec:
  containers:
    - name: nginx
      image: nginx:1.14.2
      ports:
        - containerPort: 80
```

You can check that only the pod having the label was injected with the agent sidecar:
```
kubectl get pods
NAME                                           READY   STATUS    RESTARTS   AGE
datadog-agent-cluster-agent-6dd88f969f-xm5jx   1/1     Running   0          3m6s
pod-with-label                                 2/2     Running   0          88s
pod-without-label                              1/1     Running   0          2s
```

Check the agent sidecar container that was injected, it should include the `DD_EKS_FARGATE` label set to `true`:

```
....
spec:
  containers:
  - image: nginx:1.14.2
    imagePullPolicy: IfNotPresent
    name: nginx
    ports:
    - containerPort: 80
      protocol: TCP
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-k4mhg
      readOnly: true
  - env:
    - name: DD_API_KEY
      valueFrom:
        secretKeyRef:
          key: api-key
          name: datadog-secret
    - name: DD_SITE
      value: datadoghq.com
    - name: DD_CLUSTER_NAME
    - name: DD_KUBERNETES_KUBELET_NODENAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.nodeName
    - name: DD_EKS_FARGATE
      value: "true"
    image: datadog/agent
    imagePullPolicy: IfNotPresent
    name: datadog-agent-injected
    resources:
      limits:
        cpu: 200m
        memory: 256Mi
      requests:
        cpu: 200m
        memory: 256Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-k4mhg
      readOnly: true
```

#### Advanced Configuration (Setting custom selectors)
```
datadog:
  apiKeyExistingSecret: datadog-secret
  appKeyExistingSecret: datadog-secret
  kubelet:
    tlsVerify: false

agents:
  enabled: false

clusterAgent:
  enabled: true
  admissionController:
    enabled: true
  env:
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_ENABLED
      value: "true"
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_NAMESPACE_SELECTORS
      value: '[{"ObjectSelector": {"MatchLabels": {"injectSidecarPodLabel": "true"}}, "NamespaceSelector": {"MatchLabels": {"injectSidecarNs": "true"}}}]'`
```
- When the cluster agent starts in the default namespace, create a namespace with `injectSidecarNs` label set to `true`
```
apiVersion: v1
kind: Namespace
metadata:
  name: ns-for-qa
  labels:
    injectSidecarNs: "true"
```
- Create a datadog secret in the newly created namespace: `kubectl create secret generic datadog-secret --from-literal api-key=$DD_API_KEY --from-literal app-key=$DD_APP_KEY  -n ns-for-qa`
- Create 4 pods:
  - one in the `default` namespace, without the sidecar injection label
```
apiVersion: v1
kind: Pod
metadata:
  name: nginx-default-no-label
  labels:
    admission.datadoghq.com/enabled: "false"
spec:
  containers:
    - name: nginx
      image: nginx:1.14.2
      ports:
        - containerPort: 80
```
  - one in the `default` namespace, having the sidecar injection label
```
apiVersion: v1
kind: Pod
metadata:
  name: nginx-default-with-label
  labels: 
    injectSidecarPodLabel: "true"
    admission.datadoghq.com/enabled: "false"
spec:
  containers:
    - name: nginx
      image: nginx:1.14.2
      ports:
        - containerPort: 80
```
  - one in the `ns-for-qa` namespace, having the sidecar injection label
```
apiVersion: v1
kind: Pod
metadata:
  namespace: ns-for-qa
  name: nginx-ns-for-qa-with-label
  labels: 
    injectSidecarPodLabel: "true"
    admission.datadoghq.com/enabled: "false"
spec:
  containers:
    - name: nginx
      image: nginx:1.14.2
      ports:
        - containerPort: 80
```
  - one in the `ns-for-qa` namespace, without the sidecar injection label
```
apiVersion: v1
kind: Pod
metadata:
  namespace: ns-for-qa
  name: nginx-ns-for-qa-without-label
  labels: 
    admission.datadoghq.com/enabled: "false"
spec:
  containers:
    - name: nginx
      image: nginx:1.14.2
      ports:
        - containerPort: 80
```

Only the pod that is created in the `ns-for-qa` namespace and having the sidecar injection label should be injected with the agent sidecar.

- Verify that only the pod created in the labeled namespace got injected with the agent sidecar.

- Further QA: Finally, you can try creating a pod with a container named `datadog-agent-injected`. This should not be injected with an agent sidecar since it already has a container with the same name. You can also try with different label selectors and namespace selectors, combining both with the same procedure described above, and verify the result is as expected.

Results:

- Pods in default namespace:
```
kubectl get pods -n default
NAME                                           READY   STATUS    RESTARTS   AGE
datadog-agent-cluster-agent-65794cb767-kq5h7   1/1     Running   0          13m
nginx-default-no-label                         1/1     Running   0          58s
nginx-default-with-label                       1/1     Running   0          51s
```
-  Pods in labeled namespace:
```
kubectl get pods -n ns-for-qa
NAME                            READY   STATUS    RESTARTS      AGE
nginx-ns-for-qa-with-label      2/2     Running   2 (15s ago)   38s
nginx-ns-for-qa-without-label   1/1     Running   0             32s
```
- The latter pod has the agent sidecar container as expected (notice that `DD_EKS_FARGATE` label was not added):
```
.....
spec:
  containers:
  - image: nginx:1.14.2
    imagePullPolicy: IfNotPresent
    name: nginx
    ports:
    - containerPort: 80
      protocol: TCP
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-wdzrc
      readOnly: true
  - env:
    - name: DD_API_KEY
      valueFrom:
        secretKeyRef:
          key: api-key
          name: datadog-secret
    - name: DD_SITE
      value: datadoghq.com
    - name: DD_CLUSTER_NAME
    - name: DD_KUBERNETES_KUBELET_NODENAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.nodeName
    image: datadog/agent
    imagePullPolicy: IfNotPresent
    name: datadog-agent-injected
    resources:
      limits:
        cpu: 200m
        memory: 256Mi
      requests:
        cpu: 200m
        memory: 256Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-wdzrc
      readOnly: true
.....

```

### Further QA:

The table below shows the expected behaviour for different configurations when the agent sidecar injection feature is enabled.

| Custom Selectors / Profiles Set | Provider Set       | Provider Supported | Expected Behaviour                                                                                                                                      |
|---------------------------------|--------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------|
| :x:                             | :heavy_check_mark: | :heavy_check_mark: | Agent sidecar should be injected on pods having the provider label key set (`agent.datadoghq.com/sidecar: <provider>`)                                  |
| :x:                             | :heavy_check_mark: | :x:                | No agent sidecar should be injected, and an error message will be logged in the cluster agent:  "agent sidecar provider is not supported: foo-provider" |
| :x:                             | :x:                | :x:                | No agent sidecar should be injected, and an error message will be logged in the cluster agent:  "agent sidecar provider is not supported"               |
| :heavy_check_mark:              | :heavy_check_mark: | :heavy_check_mark: | The agent sidecar container should be injected only on pods matching the selector, and the `DD_EKS_FARGATE` label should be set to `true`               |
| :heavy_check_mark:              | :heavy_check_mark: | :x:                | Agent sidecar should be injected, and you must find an error message in the cluster agent logs "unsupported provider: foo"                              |
| :heavy_check_mark:              | :x:                | :x:                | The agent sidecar container should be injected only on pods matching the selector                                                                       |

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by git-thuerk-done — APPROVED** [review]
good for docs

**Review by ogaca-dd — APPROVED** [review]

File: cmd/cluster-agent/subcommands/start/command.go [hunk 1]
```diff
@@ -11,6 +11,8 @@ package start
 import (
 	"context"
 	"fmt"
+
+	agentsidecar "github.com/DataDog/datadog-agent/pkg/clusteragent/admission/mutate/agent_sidecar"
```
Comment (clamoriniere) [inline cmd/cluster-agent/subcommands/start/command.go:15]: wrong import placement

File: pkg/clusteragent/admission/controllers/webhook/common.go [hunk 2]
```diff
@@ -8,13 +8,54 @@
 package webhook
 
 import (
+	"encoding/json"
+	"fmt"
```
Comment (clamoriniere) [inline pkg/clusteragent/admission/controllers/webhook/common.go:12]:
```suggestion
	"fmt"

```

File: pkg/clusteragent/admission/controllers/webhook/controller_v1beta1_test.go [hunk 3]
```diff
@@ -10,6 +10,7 @@ package webhook
 import (
 	"context"
 	"fmt"
+	admiv1 "k8s.io/api/admissionregistration/v1"
```
Comment (clamoriniere) [inline pkg/clusteragent/admission/controllers/webhook/controller_v1beta1_test.go:13]: wrong import placement

File: pkg/config/setup/config.go [hunk 4]
```diff
@@ -1069,6 +1069,12 @@ func InitConfig(config pkgconfigmodel.Config) {
 	config.BindEnvAndSetDefault("admission_controller.cws_instrumentation.image_tag", "latest")
 	config.BindEnv("admission_controller.cws_instrumentation.init_resources.cpu")
 	config.BindEnv("admission_controller.cws_instrumentation.init_resources.memory")
+	config.BindEnvAndSetDefault("admission_controller.agent_sidecar.enabled", false)
+	config.BindEnvAndSetDefault("admission_controller.agent_sidecar.endpoint", "/agentsidecar")
+	// Should be able to parse it to []map[string]string
+	config.BindEnvAndSetDefault("admission_controller.agent_sidecar.label_selectors", "[]")
+	// Should be able to parse it to []map[string]string
+	config.BindEnvAndSetDefault("admission_controller.agent_sidecar.namespace_selectors", "[]")
```
Comment (clamoriniere) [inline pkg/config/setup/config.go:1077]: I thought we said that for the first implementation we only support `labels_selectors` with the default label key `agent.datadoghq.com/sidecar`

File: pkg/clusteragent/admission/mutate/agent_sidecar/util.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/DataDog/datadog-agent/pkg/config"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+const agentSidecarContainerName = "datadog-agent-injected"
+
+func getDefaultSidecarTemplate() *corev1.Container {
+	agentContainer := &corev1.Container{
+		Env: []corev1.EnvVar{
```
Comment (clamoriniere) [inline pkg/clusteragent/admission/mutate/agent_sidecar/util.go:15]: I don't see the `DD_FARGATE` envvar? 
is it plan to add it in a following pull request? 
IMO we can add it now, because for now the only environment we want to support is Fargate and without it, the agent is not behaving correctly as a side car container

**Review by clamoriniere — COMMENTED** [review]
view comment during initial review

File: pkg/clusteragent/admission/mutate/agent_sidecar/util.go (hunk 5 above)
Comment (adel121) [reply pkg/clusteragent/admission/mutate/agent_sidecar/util.go:15]: This is intentional here.

Based on my previous discussion with @levan-m , I understand that in order to keep the solution generic, we will start by having the `DD_EKS_FARGATE` env var set via the env overrides (which will be implemented in a future PR). So the user will have to set `DD_EKS_FARGATE` to `true` in the env overrides when running in fargate.

In the future, we can improve this by having an admission-controller-level config (like `admission.agent-sidecar.runtime` to set the environment on a global level. 

@levan-m any thoughts?

File: pkg/config/setup/config.go (hunk 4 above)
Comment (adel121) [reply pkg/config/setup/config.go:1077]: I remember this was an initial thought, but after that based on the following discussions, I understood that we use namespace and label selectors and constrain user to specify only one selector.

With this, if users want to use `agent.datadoghq.com/sidecar` , they could, but they also have the flexibility to configure selectors based on namespaces (or multiple labels) for example.

File: pkg/clusteragent/admission/mutate/agent_sidecar/util.go (hunk 5 above)
Comment (levan-m) [reply pkg/clusteragent/admission/mutate/agent_sidecar/util.go:15]: Yes, we didn't include `DD_EKS_FARGATE` so the AC feature is generic and can be run e.g. on Kind. At the very least we will provide `DD_EKS_FARGATE` env var as an env override via Helm. If we have time we will add a runtime/provider setting and add the env var in binary.

File: pkg/config/setup/config.go (hunk 4 above)
Comment (levan-m) [reply pkg/config/setup/config.go:1077]: With the current scope we support generic object and namespace selectors [AC selectors](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-objectselector) to target pods (only). This way users can configure targeting in AC instead of configuring their applications.

File: pkg/clusteragent/admission/controllers/webhook/common.go [hunk 6]
```diff
@@ -8,13 +8,44 @@
...
 	"github.com/DataDog/datadog-agent/pkg/clusteragent/admission/mutate"
 	"github.com/DataDog/datadog-agent/pkg/config"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// Selector specifies an object label selector and a namespace label selector
+type Selector struct {
+	ObjectSelector    metav1.LabelSelector `yaml:"objectSelector,omitempty"`
+	NamespaceSelector metav1.LabelSelector `yaml:"nbjectSelector,omitempty"`
```
Comment (levan-m) [inline pkg/clusteragent/admission/controllers/webhook/common.go:24]:
```suggestion
	NamespaceSelector metav1.LabelSelector `yaml:"namespaceSelector,omitempty"`
```

**Review by levan-m — COMMENTED** [review]
Looks good overall. Left some comments possibly for minor changes!
Thank!

File: pkg/clusteragent/admission/mutate/agent_sidecar/util.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+				Name: "DD_KUBERNETES_KUBELET_NODENAME",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						APIVersion: "v1",
+						FieldPath:  "spec.nodeName",
+					},
+				},
+			},
+		},
+		Image:           "datadog/agent",
```
[the remark this answers is not in the document]
Comment (adel121) [reply pkg/clusteragent/admission/mutate/agent_sidecar/util.go:55]: agree

will add this in my next PR.

File: pkg/clusteragent/admission/controllers/webhook/controller_v1_test.go [hunk 8]
```diff
@@ -677,6 +677,128 @@ func TestGenerateTemplatesV1(t *testing.T) {
...
+				mockConfig.SetWithoutSource("admission_controller.mutate_unlabelled", false)
+				mockConfig.SetWithoutSource("admission_controller.namespace_selector_fallback", false)
+				mockConfig.SetWithoutSource("admission_controller.inject_config.enabled", false)
+				mockConfig.SetWithoutSource("admission_controller.inject_tags.enabled", false)
+				mockConfig.SetWithoutSource("admission_controller.auto_instrumentation.enabled", false)
+				mockConfig.SetWithoutSource("admission_controller.cws_instrumentation.enabled", false)
+				mockConfig.SetWithoutSource("admission_controller.cws_instrumentation.mutate_unlabelled", false)
+				mockConfig.SetWithoutSource("admission_controller.agent_sidecar.enabled", true)
+				mockConfig.SetWithoutSource("admission_controller.agent_sidecar.provider", "fargate")
+				mockConfig.SetWithoutSource("admission_controller.agent_sidecar.selectors", "[{\"NamespaceSelector\": {\"MatchLabels\":{\"labelKey1\": \"labelVal1\"}}} , {\"ObjectSelector\": {\"MatchLabels\": {\"labelKey2\": \"labelVal2\"}}}]")
```
[the remark this answers is not in the document]
Comment (adel121) [reply pkg/clusteragent/admission/controllers/webhook/controller_v1_test.go:785]: 🤔

File: cmd/cluster-agent/subcommands/start/command.go [hunk 9]
```diff
@@ -382,6 +384,7 @@ func start(log log.Component, config config.Component, taggerComp tagger.Compone
 			server.Register(pkgconfig.Datadog.GetString("admission_controller.inject_config.endpoint"), mutate.InjectConfig, apiCl.DynamicCl, apiCl.Cl)
 			server.Register(pkgconfig.Datadog.GetString("admission_controller.inject_tags.endpoint"), mutate.InjectTags, apiCl.DynamicCl, apiCl.Cl)
 			server.Register(pkgconfig.Datadog.GetString("admission_controller.auto_instrumentation.endpoint"), mutate.InjectAutoInstrumentation, apiCl.DynamicCl, apiCl.Cl)
+			server.Register(pkgconfig.Datadog.GetString("admission_controller.agent_sidecar.endpoint"), agentsidecar.InjectAgentSidecar, apiCl.DynamicCl, apiCl.Cl)
```
[the remark this answers is not in the document]
Comment (adel121) [reply cmd/cluster-agent/subcommands/start/command.go:387]: this can be a solution.

however, I think this can be dealt with later. The change can (and should) be transparent to the end user.

File: pkg/clusteragent/admission/mutate/agent_sidecar/util.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+						Key: "api-key",
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: "datadog-secret",
+						},
+					},
+				},
+			},
+			{
+				Name:  "DD_SITE",
+				Value: "datadoghq.com",
```
[the remark this answers is not in the document]
Comment (adel121) [reply pkg/clusteragent/admission/mutate/agent_sidecar/util.go:39]: agree

File: pkg/clusteragent/admission/controllers/webhook/common.go [hunk 11]
```diff
@@ -8,13 +8,54 @@
...
+func buildAgentSidecarObjectSelectors() (namespaceSelector, objectSelector *metav1.LabelSelector) {
+
+	// Read and parse label selectors
+	labelSelectorsJSON := config.Datadog.GetString("admission_controller.agent_sidecar.label_selectors")
+
+	var labelSelectors []map[string]string
+
+	err := json.Unmarshal([]byte(labelSelectorsJSON), &labelSelectors)
+	if err != nil {
+		fmt.Printf("failed to parse label selectors for admission controller agent sidecar injection webhook: %s", err)
```
Comment (GustavoCaso) [inline pkg/clusteragent/admission/controllers/webhook/common.go]: Should we be logging rather than using `fmt` ?

**Review by GustavoCaso — APPROVED** [review]
LGTM!

File: pkg/clusteragent/admission/controllers/webhook/common.go [hunk 12]
```diff
@@ -8,13 +8,59 @@
...
 )
 
+// Selector specifies an object label selector and a namespace label selector
+type Selector struct {
+	ObjectSelector    metav1.LabelSelector `yaml:"objectSelector,omitempty"`
+	NamespaceSelector metav1.LabelSelector `yaml:"namespaceSelector,omitempty"`
+}
+
+// buildAgentSidecarObjectSelectors returns the mutating webhooks object selectors based on the configuration
+func buildAgentSidecarObjectSelectors() (namespaceSelector, objectSelector *metav1.LabelSelector) {
```
Comment (davidor) [inline pkg/clusteragent/admission/controllers/webhook/common.go:29]: This is specific for the agent sidecar, shouldn't it be defined somewhere in the new `agent_sidecar` package instead of this `common` file ?

Comment (adel121) [reply pkg/clusteragent/admission/controllers/webhook/common.go:29]: I definitely agree with you 👍 

However, I am following the existing pattern.

For example, `buildCWSInstrumentationLabelSelectors` is specific to CWS Instrumentation, yet it is defined in this `common.go` file.

I thought it would be good to stick with the existing pattern until the admission controller code is refactored to follow a structured organisation.

WDYT?

Comment (davidor) [reply pkg/clusteragent/admission/controllers/webhook/common.go:29]: We can leave it for later :+1:

That would be part of the refactor I'm working on now.

File: pkg/clusteragent/admission/controllers/webhook/common.go [hunk 13]
```diff
@@ -8,13 +8,59 @@
...
+
+	var selectors []Selector
+
+	err := json.Unmarshal([]byte(selectorsJSON), &selectors)
+	if err != nil {
+		log.Errorf("failed to parse selectors for admission controller agent sidecar injection webhook: %s", err)
+		return nil, nil
+	}
+
+	if len(selectors) > 0 {
```
[the remark this answers is not in the document]
Comment (levan-m) [reply pkg/clusteragent/admission/controllers/webhook/common.go:41]: This is how I see it now, we haven't agreed yet how this will be extended some take this with some grain of salt.

When customers upgrade to the next version of sidecar injection we will have to make fewer assumptions about existing configs, i.e. we will always have one selector, one override at most. In that case we would handle this 'special case' by initializing new fields in the structs with some defaults e.g. `selector.name: default`, `profile.selectors: [default]` and new code would just handle it as a normal case. If we have multiple selectors or profiles - that wouldn't would. 
Forcing smaller input space (without limiting functionality) will make the change easier in the future.

[the remark this answers is not in the document]
Comment (adel121) [reply pkg/clusteragent/admission/controllers/webhook/common.go]: Makes sense

PR updated

Thanks

File: pkg/clusteragent/admission/controllers/webhook/common.go [hunk 14]
```diff
@@ -8,13 +8,54 @@
...
+
+	err = json.Unmarshal([]byte(nsSelectorsJSON), &nsSelectors)
+	if err != nil {
+		fmt.Printf("failed to parse namespace selectors for admission controller agent sidecar injection webhook: %s", err)
+		return nil, nil
+	}
+
+	if len(labelSelectors) > 0 {
+		objectSelector = &metav1.LabelSelector{
+			MatchLabels: labelSelectors[0],
```
Comment (davidor) [inline pkg/clusteragent/admission/controllers/webhook/common.go]: Why only the first one?

File: comp/core/hostname/hostnameinterface/go.sum [hunk 15]
```diff
@@ -23,6 +23,7 @@ go.uber.org/atomic v1.7.0 h1:ADUqmZGgLDDfbSL9ZmPxKTybcoEYHgpYfELNoN+7hsw=
 go.uber.org/atomic v1.7.0/go.mod h1:fEN4uk6kAWBTFdckzkM89CLk9XfWZrxpCo0nPH17wJc=
 go.uber.org/dig v1.17.0 h1:5Chju+tUvcC+N7N6EV08BJz41UZuO3BmHcN4A287ZLI=
 go.uber.org/dig v1.17.0/go.mod h1:rTxpf7l5I0eBTlE6/9RL+lDybC7WFwY2QH55ZSjy1mU=
+go.uber.org/fx v1.18.2/go.mod h1:g0V1KMQ66zIRk8bLu3Ea5Jt2w/cHlOIp4wdRsgh0JaY=
```
Comment (davidor) [inline comp/core/hostname/hostnameinterface/go.sum:26]: This doesn't seem to be related to this commit.
I don't think this is an issue, but mentioning it just in case.

File: pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+	for _, container := range pod.Spec.Containers {
+		if container.Name == agentSidecarContainerName {
+			log.Info("skipping agent sidecar injection: agent sidecar already exists")
+			return nil
+		}
+	}
+
+	agentSidecarContainer := getDefaultSidecarTemplate()
```
Comment (davidor) [inline pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go:38]: Not a blocker, but the `getDefaultSidecarTemplate` func is only called from here. I think it could be defined below. Specially considering that the other file doesn't have a descriptive name (`util.go`).

File: pkg/clusteragent/admission/mutate/agent_sidecar/util.go [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+			Requests: map[corev1.ResourceName]resource.Quantity{
+				"memory": resource.MustParse("256Mi"),
+				"cpu":    resource.MustParse("200m"),
+			},
+			Limits: map[corev1.ResourceName]resource.Quantity{
+				"memory": resource.MustParse("256Mi"),
+				"cpu":    resource.MustParse("200m"),
+			},
+		},
+	}
```
Comment (davidor) [inline pkg/clusteragent/admission/mutate/agent_sidecar/util.go:74]: I wonder if the resources should be configurable. If in a specific env these are not enough there's no way to change them.

File: pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar_test.go [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+
+			err := injectAgentSidecar(test.Pod, "", nil)
+
+			if test.ExpectError {
+				assert.Error(tt, err, "expected non-nil error to be returned")
+			} else {
+				assert.NoError(tt, err, "expected returned error to be nil")
+
+				if test.ShouldInject {
+					assert.Equalf(tt, len(test.Pod.Spec.Containers), containersCount+1, "should inject sidecar")
```
Comment (davidor) [inline pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar_test.go]: Should we at least validate the image name of the new container?

File: pkg/clusteragent/admission/mutate/agent_sidecar/README.md [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+| :x:                             | :heavy_check_mark: | :heavy_check_mark: | Agent sidecar should be injected on pods having the provider label key set (`agent.datadoghq.com/sidecar: <provider>`)                                  |
+| :x:                             | :heavy_check_mark: | :x:                | No agent sidecar should be injected, and an error message will be logged in the cluster agent:  "agent sidecar provider is not supported: foo-provider" |
+| :x:                             | :x:                | :x:                | No agent sidecar should be injected, and an error message will be logged in the cluster agent:  "agent sidecar provider is not supported"               |
+| :heavy_check_mark:              | :heavy_check_mark: | :heavy_check_mark: | The agent sidecar container should be injected only on pods matching the selector, and the `DD_EKS_FARGATE` label should be set to `true`               |
+| :heavy_check_mark:              | :heavy_check_mark: | :x:                | Agent sidecar should be injected, and you must find an error message in the cluster agent logs "unsupported provider: foo"                              |
```
Comment (davidor) [inline pkg/clusteragent/admission/mutate/agent_sidecar/README.md:78]: shouldn't be injected

File: pkg/clusteragent/admission/mutate/agent_sidecar/README.md [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+| :heavy_check_mark:              | :x:                | :x:                | The agent sidecar container should be injected only on pods matching the selector                                                                       |
+
+
+
+
+## Notes
+- For now, we only support configuring 1 custom selector and 1 custom profile.
+- For now, only `fargate` provider is supported
+- For now, only 1 selector and 1 profile (config override) can be configured.
```
Comment (davidor) [inline pkg/clusteragent/admission/mutate/agent_sidecar/README.md:87]: Might be useful to add an example of how to define the custom selector.

File: pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go (hunk 16 above)
Comment (davidor) [reply pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go:38]: I think it'd move `getDefaultSidecarTemplate` here and rename `util.go` to `provider.go`.

File: pkg/clusteragent/admission/controllers/webhook/common.go (hunk 14 above)
Comment (adel121) [reply pkg/clusteragent/admission/controllers/webhook/common.go]: In the first version of this feature, we allow configuring at most 1 custom selector.

In a future version, we will allow setting several selectors and configuring overrides for each of them. 

We are defining the selector config as a slice of selectors in order not to break user setups in the future when we upgrade to version 2, while constraining selectors to only 1 selector in version 1.

File: pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go (hunk 16 above)
Comment (adel121) [reply pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar.go:38]: I agree. 👍

File: pkg/clusteragent/admission/mutate/agent_sidecar/util.go (hunk 17 above)
Comment (adel121) [reply pkg/clusteragent/admission/mutate/agent_sidecar/util.go:74]: Yes resources are configurable. 

I have another [PR](https://github.com/DataDog/datadog-agent/pull/22854) that I opened to allow customising env vars and resource limits.

File: pkg/clusteragent/admission/mutate/agent_sidecar/README.md (hunk 19 above)
Comment (adel121) [reply pkg/clusteragent/admission/mutate/agent_sidecar/README.md:78]: Why do you think so?

If user set a custom selector, pods matching the selector should be injected.

Having an unsupported provider should lead to failure in setting provider-specific overrides. But why should it lead to failure of injection?

File: pkg/clusteragent/admission/mutate/agent_sidecar/README.md (hunk 20 above)
Comment (adel121) [reply pkg/clusteragent/admission/mutate/agent_sidecar/README.md:87]: Yes indeed.

Currently we still didn't add configs for this feature in helm and operator. Showing an example here would require configuring environment variables (as you can see in the QA), which is not very nice to have.

I think it would be better to wait until @levan-m updates helm and the operator to support configuring this feature easily and then we can update the readme here.

File: pkg/clusteragent/admission/mutate/agent_sidecar/README.md (hunk 19 above)
Comment (davidor) [reply pkg/clusteragent/admission/mutate/agent_sidecar/README.md:78]: OK. I think it'll make more sense once there are more providers :+1:

**Review by davidor — APPROVED** [review]

**Review by levan-m — APPROVED** [review]
